### DISCUSSION TITLE
Power Query Integration

### DISCUSSION THREAD

**Discussion by JeridInsight** [TabularEditor/TabularEditor3#1135] [Ideas]
I asked Miguel Escobar this question at the Microsoft Fabric Engineering Connection call:

" What would it take to integrate Power Query into Tabular Editor? It would be really great if you could use it right inside of TE3 for building your M query for partitions and previewing the data. Currently you have to prototype your query in PBI Desktop and then paste it over into TE3. Is it a licensing issue or just a matter of Kapacity integrating it into their product? ".

 His response was that you could leverage the [Power Query SDK](https://learn.microsoft.com/en-us/power-query/install-sdk).
 
 This would be really nice!!!

**Comment by otykier** [comment]
I’m afraid the Power Query SDK is intended for creating custom connectors using Power Query. From what I can tell, it does not in any way bring Power Query functionality into 3rd party applications, unfortunately.

**Reply by JeridInsight** [reply]
That sucks.  I wonder why Miguel said that then?  He's very deeply involved in Power Query and Azure Data Factory and I assumed he would understand what I was asking.  I mean what I provided above is exactly what I asked him.  

Oh well.  I just thought I would ask them for you...  : )    I really think that would be a nice addition to your tool

**Reply by otykier** [reply]
I'm sure Miguel just misunderstood the question. I've actually asked him the same thing several years ago, but no luck.

The good news is that TE3 already has an M / Power Query code parser, so we can definitely improve the editing experience, adding things like auto-complete, syntax checking, auto-formatting, etc. (just like we have for DAX editing). But we will unfortunately not be able to recreate the full Power Query experience with UI for previewing data, applying transformations, etc.

**Reply by GreyskullAnalytics** [reply]
You can use the SDK as a kind of M client, so you can write M and preview your results. Maybe he was suggesting an alternative to using PBI desktop?
